src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -974,7 +979,16 @@ impl<'a, 'tcx> Visitor<'tcx> for Qualifier<'a, 'tcx, 'tcx> {
                     if !constant_arguments.contains(&i) {
                         return
                     }
-                    if this.qualif.is_empty() {
+                    // Since the argument is required to be constant,
+                    // we care about constness, not promotability.
+                    // If we checked for promotability, we'd miss out on
+                    // the results of function calls (which are never promoted)
+                    // This is not a problem, because the argument explicitly
+                    // requests constness, in contrast to regular promotion
+                    // which happens even without the user requesting it.
+                    // We can error out with a hard error if the argument is not
+                    // constant here.
+                    if (this.qualif - Qualif::NOT_PROMOTABLE).is_empty() {
```

Comment:
Thanks for expanding on this. I still think the terminology is a bit confusing: it seems to suggest "return values are never promotable but we're going to promote one here anyway, because it's const."

Reply:
Well... yes. The rules for automatic promotion don't include function calls. The rules for constants do include function calls. Since the current function requires a constant here, if we did something like `0usize - 1` we'd get an error. If you just do `&(0usize - 1)` in normal runtime code, you get a warning and a panic at runtime (in debug mode).

Reply:
Yeah sure, I agree this makes sense, but could we just change the comment to "never promoted for runtime-evaluated/non-const functions" or such, just to be super-clear and make me happy? :-)